crates/trie/parallel/src/proof_task.rs

```diff
@@ -0,0 +1,356 @@
+    /// A receiver for new proof tasks.
+    proof_task_rx: Receiver<ProofTaskMessage>,
+}
+
+impl<Factory: DatabaseProviderFactory> ProofTaskManager<Factory> {
+    /// Creates a new [`ProofTaskManager`] with the given max concurrency, creating that number of
+    /// cursor factories.
+    ///
+    /// Returns an error if the consistent view provider fails to create a read-only transaction.
+    pub fn new(
+        executor: Handle,
+        view: ConsistentDbView<Factory>,
+        task_ctx: ProofTaskCtx,
+        max_concurrency: usize,
+        proof_task_rx: Receiver<ProofTaskMessage>,
+    ) -> Self {
+        Self {
+            max_concurrency,
+            total_transactions: 0,
+            view,
+            task_ctx,
+            pending_proofs: VecDeque::new(),
+            executor,
+            proof_task_txs: Vec::new(),
+            running_proof_tasks: Vec::new(),
+            proof_task_rx,
+        }
+    }
+}
+
+impl<Factory> ProofTaskManager<Factory>
+where
+    Factory: DatabaseProviderFactory<Provider: BlockReader> + StateCommitmentProvider + 'static,
+{
+    /// Spawns the proof task on the executor, with the input multiproof targets.
+    ///
+    /// If a task cannot be spawned immediately, this will be queued for completion later.
+    pub fn queue_proof_task(
+        &mut self,
+        input: StorageProofInput,
+        sender: Sender<StorageProofResult>,
+    ) {
+        self.pending_proofs.push_back((input, sender));
+    }
+
+    /// Gets either the next available transaction, or creates a new one if all are in use and the
+    /// total number of transactions created is less than the max concurrency.
+    pub fn get_or_create_tx(&mut self) -> ProviderResult<Option<ProofTaskTx<FactoryTx<Factory>>>> {
+        if let Some(proof_task_tx) = self.proof_task_txs.pop() {
+            return Ok(Some(proof_task_tx));
+        }
+
+        // if we can create a new tx within our concurrency limits, create one on-demand
+        if self.total_transactions < self.max_concurrency {
+            let provider_ro = self.view.provider_ro()?;
+            let tx = provider_ro.into_tx();
+            self.total_transactions += 1;
+            return Ok(Some(ProofTaskTx::new(tx, self.task_ctx.clone())));
+        }
+
+        Ok(None)
+    }
+
+    /// Spawns the next queued proof task on the executor with the given input, if there are any
+    /// transactions available.
+    ///
+    /// This will return an error if a transaction must be created on-demand and the consistent view
+    /// provider fails.
+    pub fn try_spawn_next(&mut self) -> ProviderResult<()> {
+        let Some(proof_task_tx) = self.get_or_create_tx()? else { return Ok(()) };
+
+        let Some((pending_proof, sender)) = self.pending_proofs.pop_front() else {
+            // if there are no targets to do anything with put the tx back
+            self.proof_task_txs.push(proof_task_tx);
+            return Ok(())
+        };
```

Comment:
should we swap these, so that txs aren't created if they won't be used?

Reply:
yeah that would be better